[spec] Add security and privacy considerations sections (#34)
WICG/scheduling-apis: spec/security.md

Security Considerations {#sec-security}
=====================

<div class="non-normative">
*This section is non-normative.*

The main security consideration for the APIs defined in this specification is
whether or not any information is potentially leaked between origins by
timing-based side-channel attacks.
</div>

`postTask` as a High-Resolution Timing Source {#sec-security-high-res-timer}
---------------------

<div class="non-normative">
The first consideration is whether the {{Scheduler/postTask()}} API can be used
as a high-resolution timing source. {{Scheduler/postTask()}}'s
{{SchedulerPostTaskOptions/delay}}, like
{{WindowOrWorkerGlobalScope/setTimeout()}}'s timeout value, is expressed in
whole milliseconds (the minimum non-zero delay being 1 ms), and there is no
guarantee that tasks will run exactly when the delay expires since tasks are
queued when the delay expires. Given this, we do not believe the API can be

> used as a high-resolution timing source, but we mention it here because of the

Reviewer: Try to lead with the lede. So something like,

> This API cannot be used as a high-resolution timing source. Like setTimeout's timeout value, postTask()'s delay is expressed in whole milliseconds (the minimum non-zero delay is 1 ms), so callers can't express any timing more precise than 1ms. Further, the fact that tasks are only queued when their delay expires, and not run instantly, further reduces the precision available to callers.

That also gives future API tweakers a clue to look closely at attempts to change those two qualities.

> used as a high-resolution timing source, but we mention it here because of the
> general interest in the topic.

Reviewer: ```suggestion
used as a high-resolution timing source.
```

I don't think the last clause adds anything for readers.

</div>

Monitoring Another Origin's Tasks {#sec-security-monitoring-tasks}
---------------------

<div class="non-normative">
The second consideration is whether {{Scheduler/postTask()}} leaks any
information about other origins' tasks. The threat model we consider is code

> from two different origins running in separate event loops in the same thread.

Reviewer: It's not quite right to call this the "threat model". Maybe just "We consider an attacker running on one origin trying to obtain information about code executing in another origin (and hence in a separate event loop) that is scheduled in the same thread in a browser."

> Because the UA can only run tasks from one event loop at a time, an attacker

Reviewer: "a thread within the UA"?

might be able to gain information about tasks running in another event loop by
monitoring when their tasks run. For example, an attacker could flood the
system with {{TaskPriority/user-blocking}} tasks and expect them to run
consecutively; if there are large gaps in between, then the attacker might
infer something about tasks running in another event loop, e.g. they have

> higher priority.

Reviewer: What does it mean for a task to have higher priority than a user-blocking one? Is it just that the attacker learns that the other colocated origin had some work to do, and the UA decided to give it a chance to run?

Author: That's what I was intending it to mean. I'll try to rewrite this so it's clearer. The problem is, as I mention below, whatever information is gained is implementation-dependent, so anything here will be hypothetical. And realistically, the information gained is likely to be probabilistic. I'm not sure the best way to express all of that in this section.

The first thing we note is that the attacker would not be able to definitively
tell that a task ran in another event loop, e.g. an internal browser task might
have run, or perhaps the UA throttled the attacker's script.  Second, any

> information gained would be implementation-dependent since inter-event-loop

Reviewer: I'm not sure it's really a defense that the attacker would have to implement their attack for each of ~3 implementations.

Reviewer: Maybe "The information exposed would depend on implementation details, and implementations can reduce it as described below."

task selection is not specified. Our opinion is that information gained in such

> an attack is likely to be benign, but there are mitigations that implementers

Reviewer: Have any of the xsleaks folks looked at this? They've been pretty clever about making leaks non-benign.

Author: I removed this statement and rewrote this section, attempting to explain what the status quo is and what the delta we're introducing is --- without making any kind of assertions of the severity.

> an attack is likely to be benign, but there are mitigations that implementers
> can consider to minimize the risk:

Reviewer: ```suggestion
an attack is likely to be benign, but implementers can use the following mitigations to minimize the risk:
```

Is Chrome implementing these? e.g. I think all the site isolation work and out-of-process iframes implement the first.

Reviewer: \o/
